cosmwasm/contracts/wormchain-accounting/src/contract.rs

```diff
@@ -0,0 +1,491 @@
+                source: ConversionOverflowError::new("u32", "usize", q.to_string()),
+            })
+        })
+        .context("failed to calculate quorum")?;
+
+    let observations: Vec<Observation> =
+        from_binary(&observations).context("failed to parse `Observations`")?;
+
+    let events = observations
+        .into_iter()
+        .map(|o| {
+            handle_observation(
+                deps.branch(),
+                o,
+                observer_set_index,
+                quorum,
+                signature.clone(),
+            )
+        })
+        .filter_map(Result::transpose)
+        .collect::<anyhow::Result<Vec<_>>>()
+        .context("failed to handle `Observation`")?;
+
+    Ok(Response::new()
+        .add_attribute("action", "submit_observations")
+        .add_attribute("owner", info.sender)
+        .add_events(events))
+}
+
+fn handle_observation(
+    mut deps: DepsMut<WormholeQuery>,
+    o: Observation,
+    observer_set_index: u32,
+    quorum: usize,
+    sig: wormhole_bindings::Signature,
+) -> anyhow::Result<Option<Event>> {
+    if accounting::has_transfer(deps.as_ref(), o.key.clone()) {
+        bail!("transfer for key \"{}\" already committed", o.key);
+    }
+
+    let key = PENDING_TRANSFERS.key(o.key.clone());
+    let mut pending = key
+        .may_load(deps.storage)
+        .map(Option::unwrap_or_default)
+        .context("failed to load `PendingTransfer`")?;
+    let data = match pending
+        .iter_mut()
+        .find(|d| d.observer_set_index() == observer_set_index && d.observation() == &o)
+    {
+        Some(d) => d,
+        None => {
+            pending.push(Data::new(o.clone(), observer_set_index));
+            let back = pending.len() - 1;
+            &mut pending[back]
+        }
+    };
+
+    if !data.add_signature(sig) {
+        bail!("cannot submit duplicate signatures for the same observation");
+    }
+
+    if data.signatures().len() < quorum {
+        // Still need more signatures so just save the pending transfer data and exit.
+        key.save(deps.storage, &pending)
+            .context("failed to save pending transfers")?;
+
+        return Ok(None);
+    }
+
+    let (msg, _) = serde_wormhole::from_slice_with_payload(&o.payload)
+        .context("failed to parse observation payload")?;
+    let tx_data = match msg {
+        Message::Transfer {
+            amount,
+            token_address,
+            token_chain,
+            recipient_chain,
+            ..
+        }
+        | Message::TransferWithPayload {
+            amount,
+            token_address,
+            token_chain,
+            recipient_chain,
+            ..
+        } => transfer::Data {
+            amount: Uint256::from_be_bytes(amount.0),
+            token_address: TokenAddress::new(token_address.0),
+            token_chain: token_chain.into(),
+            recipient_chain: recipient_chain.into(),
+        },
+        _ => bail!("Unknown tokenbridge payload"),
```

Comment:
It might be nice to return this error immediately rather than collecting 13 signatures for it and then failing.

Reply:
Same comment as above.  This is an optimization with the assumption that observations that reach quorum are highly unlikely to have invalid or unsupported payloads.